### DISCUSSION TITLE
Document when to use --clean

### DISCUSSION THREAD

**Discussion by Thoufak** [pyinstaller/pyinstaller#7116] [PyInstaller] [answered]
There's a documentation about the --clean option [here](https://pyinstaller.org/en/stable/usage.html#cmdoption-clean), however it doesn't make it clear when one should use it. To be honest, I don't understand at all why it's useful. Can the mentioned cache and temporary files make any harm? Can you get any kind of inconsistencies if you don't remove them? I'd like to suggest adding more information to the description of this option.

**Comment by bwoodsend** [comment, marked as answer]
Yeah, the `--clean` situation is a bit of a mess. Just from memory, here are the main cases:

The following cases don't need a full rescan and rebuild and PyInstaller correctly caches where appropriate. Using `--clean` here will slow down rebuilds.

* When you add/change a `--add-data`/`--collect-data-files` option or their equivalents in the spec file
* You update a data file or icon's contents
* You rebuild without changing anything

It has to do a full rescan of all your imports in the cases below. PyInstaller again does this correctly. Using `--clean` will not make any difference because PyInstaller does it already.

* Any of your `.py` files change. Note that there is some funkiness related to detection of changes to `.py` files in dependencies which I've never quite pinned down 
* You change any of the import related options (`--hiddenimport`, `--collect-submodules`, `--excludes`, etc)

It fails to detect and mistakenly caches (i.e. you must use `--clean` or it will refuse to update) in any of the following:

* You install/upgrade/reinstall one of your dependencies
* You install/upgrade/reinstall PyInstaller or pyinstaller-hooks-contrib
* You create/add/modify a hook file

**Reply by Thoufak** [reply]
Thank you for a detailed response.

I think this will be enough for the docs:

> (you must use --clean or it will refuse to update) in any of the following:
> 
> - You install/upgrade/reinstall one of your dependencies
> - You install/upgrade/reinstall PyInstaller or pyinstaller-hooks-contrib
> - You create/add/modify a hook file

**Reply by Thoufak** [reply]
Also, may I know where these cache and temporary files are stored? I'd like to clear them manually.

**Reply by rokm** [reply]
> Also, may I know where these cache and temporary files are stored? I'd like to clear them manually.

The default location on linux is either `$XDG_DATA_HOME/pyinstaller` or `$XDG_CACHE_HOME/pyinstaller` depending on which environment variable is defined; on my Fedora system, this ends up using `XDG_CACHE_HOME` and translates into `~/.cache/pyinstaller.

On Windows, it should be `%LOCALAPPDATA%\pyinstaller`.

That is, unless you override the location with `PYINSTALLER_CONFIG_DIR`.

For details, see [here](https://github.com/pyinstaller/pyinstaller/blob/b3087846a35a799d60b17c85cce480f42dcedb55/PyInstaller/configure.py#L60-L87).

The clean option actually clears both the (global) cache and the build-specific "workdir" ([here](https://github.com/pyinstaller/pyinstaller/blob/b3087846a35a799d60b17c85cce480f42dcedb55/PyInstaller/building/build_main.py#L833)); the latter being the `build` directory that PyInstaller typically places next to your spec file (which contains intermediate build results that may be reused).
